// @nomicfoundation/hardhat-ignition/dist/src/helpers/pretty-event-handler.js
import { DeploymentResultType, ExecutionEventResultType, } from "@nomicfoundation/ignition-core";
import { calculateBatchDisplay } from "../internal/ui/helpers/calculate-batch-display.js";
import { calculateDeployingModulePanel } from "../internal/ui/helpers/calculate-deploying-module-panel.js";
import { calculateDeploymentCompleteDisplay } from "../internal/ui/helpers/calculate-deployment-complete-display.js";
import { calculateStartingMessage } from "../internal/ui/helpers/calculate-starting-message.js";
import { wasAnythingExecuted } from "../internal/ui/helpers/was-anything-executed.js";
import { UiFutureStatusType, UiStateDeploymentStatus, } from "../internal/ui/types.js";
export class PrettyEventHandler {
    _interruptions;
    _config;
    _uiState = {
        status: UiStateDeploymentStatus.UNSTARTED,
        chainId: null,
        moduleName: null,
        deploymentDir: null,
        batches: [],
        currentBatch: 0,
        result: null,
        warnings: [],
        isResumed: null,
        maxFeeBumps: 0,
        disableFeeBumping: null,
        gasBumps: {},
        strategy: null,
    };
    constructor(_interruptions, config) {
        this._interruptions = _interruptions;
        this._config = {
            ...{
                deploymentParams: {},
                disableOutput: false,
            },
            ...config,
        };
    }
    get state() {
        return this._uiState;
    }
    set state(uiState) {
        this._uiState = uiState;
    }
    async deploymentStart(event) {
        this.state = {
            ...this.state,
            status: UiStateDeploymentStatus.DEPLOYING,
            moduleName: event.moduleName,
            deploymentDir: event.deploymentDir,
            isResumed: event.isResumed,
            maxFeeBumps: event.maxFeeBumps,
            disableFeeBumping: event.disableFeeBumping,
        };
        await this._writeToDisplay(DisplayStage.STARTING_DEPLOYMENT);
    }
    async deploymentInitialize(event) {
        this.state = {
            ...this.state,
            chainId: event.chainId,
        };
    }
    async runStart(_event) {
        await this._writeToDisplay(DisplayStage.RUN_START);
    }
    async beginNextBatch(_event) {
        // rerender the previous batch
        if (this.state.currentBatch > 0) {
            await this._writeToDisplay(DisplayStage.REDISPLAY_BATCH);
        }
        this.state = {
            ...this.state,
            currentBatch: this.state.currentBatch + 1,
        };
        if (this.state.currentBatch === 0) {
            return;
        }
        // render the new batch
        await this._writeToDisplay(DisplayStage.DISPLAY_NEXT_BATCH);
    }
    async wipeApply(event) {
        const batches = [];
        for (const batch of this.state.batches) {
            const futureBatch = [];
            for (const future of batch) {
                if (future.futureId === event.futureId) {
                    continue;
                }
                else {
                    futureBatch.push(future);
                }
            }
            batches.push(futureBatch);
        }
        this.state = {
            ...this.state,
            batches,
        };
    }
    async deploymentExecutionStateInitialize(event) {
        await this._setFutureStatusInitializedAndRedisplayBatch(event);
    }
    async deploymentExecutionStateComplete(event) {
        await this._setFutureStatusCompleteAndRedisplayBatch(event);
    }
    async callExecutionStateInitialize(event) {
        await this._setFutureStatusInitializedAndRedisplayBatch(event);
    }
    async callExecutionStateComplete(event) {
        await this._setFutureStatusCompleteAndRedisplayBatch(event);
    }
    async staticCallExecutionStateInitialize(event) {
        await this._setFutureStatusInitializedAndRedisplayBatch(event);
    }
    async staticCallExecutionStateComplete(event) {
        await this._setFutureStatusCompleteAndRedisplayBatch(event);
    }
    async sendDataExecutionStateInitialize(event) {
        await this._setFutureStatusInitializedAndRedisplayBatch(event);
    }
    async sendDataExecutionStateComplete(event) {
        await this._setFutureStatusCompleteAndRedisplayBatch(event);
    }
    async contractAtExecutionStateInitialize(event) {
        await this._setFutureStatusAndRedisplayBatch(event.futureId, {
            type: UiFutureStatusType.SUCCESS,
        });
    }
    async readEventArgumentExecutionStateInitialize(event) {
        await this._setFutureStatusAndRedisplayBatch(event.futureId, {
            type: UiFutureStatusType.SUCCESS,
        });
    }
    async encodeFunctionCallExecutionStateInitialize(event) {
        await this._setFutureStatusAndRedisplayBatch(event.futureId, {
            type: UiFutureStatusType.SUCCESS,
        });
    }
    async batchInitialize(event) {
        const batches = [];
        for (const batch of event.batches) {
            const futureBatch = [];
            for (const futureId of batch) {
                futureBatch.push({
                    futureId,
                    status: {
                        type: UiFutureStatusType.UNSTARTED,
                    },
                });
            }
            batches.push(futureBatch);
        }
        this.state = {
            ...this.state,
            batches,
        };
    }
    async networkInteractionRequest(_event) { }
    async transactionPrepareSend(_event) { }
    async transactionSend(_event) { }
    async transactionConfirm(_event) { }
    async staticCallComplete(_event) { }
    async onchainInteractionBumpFees(event) {
        if (this._uiState.gasBumps[event.futureId] === undefined) {
            this._uiState.gasBumps[event.futureId] = 0;
        }
        this._uiState.gasBumps[event.futureId] += 1;
        await this._writeToDisplay(DisplayStage.REDISPLAY_BATCH);
    }
    async onchainInteractionDropped(_event) { }
    async onchainInteractionReplacedByUser(_event) { }
    async onchainInteractionTimeout(_event) { }
    async deploymentComplete(event) {
        this.state = {
            ...this.state,
            status: UiStateDeploymentStatus.COMPLETE,
            result: event.result,
            batches: this._applyResultToBatches(this.state.batches, event.result),
        };
        await this._writeToDisplay(DisplayStage.DEPLOYMENT_COMPLETE);
    }
    async reconciliationWarnings(event) {
        this.state = {
            ...this.state,
            warnings: [...this.state.warnings, ...event.warnings],
        };
    }
    async setModuleId(event) {
        this.state = {
            ...this.state,
            moduleName: event.moduleName,
        };
    }
    async setStrategy(event) {
        this.state = {
            ...this.state,
            strategy: event.strategy,
        };
    }
    async _setFutureStatusInitializedAndRedisplayBatch({ futureId, }) {
        await this._setFutureStatusAndRedisplayBatch(futureId, {
            type: UiFutureStatusType.UNSTARTED,
        });
    }
    async _setFutureStatusCompleteAndRedisplayBatch({ futureId, result, }) {
        await this._setFutureStatusAndRedisplayBatch(futureId, this._getFutureStatusFromEventResult(result));
        this.state = {
            ...this.state,
        };
    }
    async _setFutureStatusAndRedisplayBatch(futureId, status) {
        const updatedFuture = {
            futureId,
            status,
        };
        this.state = {
            ...this.state,
            batches: this._applyUpdateToBatchFuture(updatedFuture),
        };
        await this._writeToDisplay(DisplayStage.REDISPLAY_BATCH);
    }
    _applyUpdateToBatchFuture(updatedFuture) {
        const batches = [];
        for (const batch of this.state.batches) {
            const futureBatch = [];
            for (const future of batch) {
                if (future.futureId === updatedFuture.futureId) {
                    futureBatch.push(updatedFuture);
                }
                else {
                    futureBatch.push(future);
                }
            }
            batches.push(futureBatch);
        }
        return batches;
    }
    _getFutureStatusFromEventResult(result) {
        switch (result.type) {
            case ExecutionEventResultType.SUCCESS: {
                return {
                    type: UiFutureStatusType.SUCCESS,
                    result: result.result,
                };
            }
            case ExecutionEventResultType.ERROR: {
                return {
                    type: UiFutureStatusType.ERRORED,
                    message: result.error,
                };
            }
            case ExecutionEventResultType.HELD: {
                return {
                    type: UiFutureStatusType.HELD,
                    heldId: result.heldId,
                    reason: result.reason,
                };
            }
        }
    }
    _applyResultToBatches(batches, result) {
        const newBatches = [];
        for (const oldBatch of batches) {
            const newBatch = [];
            for (const future of oldBatch) {
                const updatedFuture = this._hasUpdatedResult(future.futureId, result);
                newBatch.push(updatedFuture ?? future);
            }
            newBatches.push(newBatch);
        }
        return newBatches;
    }
    _hasUpdatedResult(futureId, result) {
        if (result.type !== DeploymentResultType.EXECUTION_ERROR) {
            return null;
        }
        const failed = result.failed.find((f) => f.futureId === futureId);
        if (failed !== undefined) {
            const f = {
                futureId,
                status: {
                    type: UiFutureStatusType.ERRORED,
                    message: failed.error,
                },
            };
            return f;
        }
        const timedout = result.timedOut.find((f) => f.futureId === futureId);
        if (timedout !== undefined) {
            const f = {
                futureId,
                status: {
                    type: UiFutureStatusType.TIMEDOUT,
                },
            };
            return f;
        }
        return null;
    }
    async _writeToDisplay(stage) {
        await this._interruptions.uninterrupted(() => {
            if (!this._config.disableOutput) {
                switch (stage) {
                    case DisplayStage.STARTING_DEPLOYMENT: {
                        process.stdout.write(calculateStartingMessage(this.state));
                        break;
                    }
                    case DisplayStage.RUN_START: {
                        this._clearCurrentLine();
                        console.log(calculateDeployingModulePanel(this.state));
                        break;
                    }
                    case DisplayStage.DISPLAY_NEXT_BATCH: {
                        console.log(calculateBatchDisplay(this.state).text);
                        break;
                    }
                    case DisplayStage.REDISPLAY_BATCH: {
                        this._redisplayCurrentBatch();
                        break;
                    }
                    case DisplayStage.DEPLOYMENT_COMPLETE: {
                        // If batches were executed, rerender the last batch
                        if (wasAnythingExecuted(this.state)) {
                            this._redisplayCurrentBatch();
                        }
                        else {
                            // Otherwise only the completion panel will be shown so clear
                            // the Starting Ignition line.
                            this._clearCurrentLine();
                        }
                        console.log(calculateDeploymentCompleteDisplay(this.state));
                        break;
                    }
                }
            }
        });
    }
    _redisplayCurrentBatch() {
        const { height, text: batch } = calculateBatchDisplay(this.state);
        this._clearUpToHeight(height);
        console.log(batch);
    }
    _clearCurrentLine() {
        if (!process.stdout.isTTY)
            return;
        process.stdout.write("\x1b[2K"); // clear entire line
        process.stdout.write("\x1b[0G"); // move cursor to column 0
    }
    _clearUpToHeight(height) {
        if (!process.stdout.isTTY)
            return;
        if (height > 0)
            process.stdout.write(`\x1b[${height}A`); // move up N lines
        process.stdout.write("\x1b[J"); // clear from cursor down
    }
}
export var DisplayStage;
(function (DisplayStage) {
    DisplayStage["STARTING_DEPLOYMENT"] = "STARTING_DEPLOYMENT";
    DisplayStage["RUN_START"] = "RUN_START";
    DisplayStage["DISPLAY_NEXT_BATCH"] = "DISPLAY_NEXT_BATCH";
    DisplayStage["REDISPLAY_BATCH"] = "REDISPLAY_BATCH";
    DisplayStage["DEPLOYMENT_COMPLETE"] = "DEPLOYMENT_COMPLETE";
})(DisplayStage || (DisplayStage = {}));
