// hardhat/dist/src/internal/builtin-plugins/artifacts/hook-handlers/hre.js
class LazyArtifactManager {
    #artifactsPath;
    #artifactManager;
    constructor(artifactsPath) {
        this.#artifactManager = undefined;
        this.#artifactsPath = artifactsPath;
    }
    async readArtifact(contractNameOrFullyQualifiedName) {
        const artifactManager = await this.#getArtifactManager();
        return await artifactManager.readArtifact(contractNameOrFullyQualifiedName);
    }
    async getArtifactPath(contractNameOrFullyQualifiedName) {
        const artifactManager = await this.#getArtifactManager();
        return await artifactManager.getArtifactPath(contractNameOrFullyQualifiedName);
    }
    async artifactExists(contractNameOrFullyQualifiedName) {
        const artifactManager = await this.#getArtifactManager();
        return await artifactManager.artifactExists(contractNameOrFullyQualifiedName);
    }
    async getAllFullyQualifiedNames() {
        const artifactManager = await this.#getArtifactManager();
        return await artifactManager.getAllFullyQualifiedNames();
    }
    async getAllArtifactPaths() {
        const artifactManager = await this.#getArtifactManager();
        return await artifactManager.getAllArtifactPaths();
    }
    async getBuildInfoId(contractNameOrFullyQualifiedName) {
        const artifactManager = await this.#getArtifactManager();
        return await artifactManager.getBuildInfoId(contractNameOrFullyQualifiedName);
    }
    async getAllBuildInfoIds() {
        const artifactManager = await this.#getArtifactManager();
        return await artifactManager.getAllBuildInfoIds();
    }
    async getBuildInfoPath(buildInfoId) {
        const artifactManager = await this.#getArtifactManager();
        return await artifactManager.getBuildInfoPath(buildInfoId);
    }
    async getBuildInfoOutputPath(buildInfoId) {
        const artifactManager = await this.#getArtifactManager();
        return await artifactManager.getBuildInfoOutputPath(buildInfoId);
    }
    async clearCache() {
        const artifactManager = await this.#getArtifactManager();
        return await artifactManager.clearCache();
    }
    async #getArtifactManager() {
        if (this.#artifactManager === undefined) {
            const { ArtifactManagerImplementation } = await import("../artifact-manager.js");
            this.#artifactManager = new ArtifactManagerImplementation(this.#artifactsPath);
        }
        return this.#artifactManager;
    }
}
export default async () => {
    const handlers = {
        created: async (_context, hre) => {
            hre.artifacts = new LazyArtifactManager(hre.config.paths.artifacts);
        },
    };
    return handlers;
};
